rcl/resource/logging_macros.h.em

```diff
@@ -0,0 +1,421 @@
+ */
+# define RCL_LIKELY(x) (x)
+/**
+ * \def RCL_UNLIKELY
+ * No op since Windows doesn't support providing branch prediction information.
+ */
+# define RCL_UNLIKELY(x) (x)
+#endif
+
+/**
+ * \def RCL_LOGGING_AUTOINIT
+ * \brief Initialize the rcl logging library.
+ * Usually it is unnecessary to call the macro directly.
+ * All logging macros ensure that this has been called once.
+ */
+#define RCL_LOGGING_AUTOINIT \
+  if (RCL_UNLIKELY(!g_rcl_logging_initialized)) { \
+    rcl_logging_initialize(); \
+  }
+
+/**
+ * \def RCL_LOG_COND_NAMED
+ * The logging macro all other logging macros call directly or indirectly.
+ * \param severity The severity level
+ * \param condition_before The condition macro(s) inserted before the log call
+ * \param condition_after The condition macro(s) inserted after the log call
+ * \param name The name of the logger
+ * \param format The format string
+ * \param ... The variable arguments for the format strings
+ */
+#define RCL_LOG_COND_NAMED(severity, condition_before, condition_after, name, format, ...) \
+  { \
+    RCL_LOGGING_AUTOINIT \
+    static RclLogLocation __rcl_logging_location = {__func__, __FILE__, __LINE__}; \
+    condition_before \
+    if (severity >= g_rcl_logging_severity_threshold) { \
+      rcl_log(&__rcl_logging_location, severity, name, format, ## __VA_ARGS__); \
+    } \
+    condition_after \
+  }
+
+///@@{
+/**
+ * \def RCL_LOG_CONDITION_EMPTY
+ * An empty macro which can be used as a placeholder for `condition_before`
+ * and `condition_after` which doesn't affect the logging call.
+ */
+#define RCL_LOG_CONDITION_EMPTY
+///@@}
+
+/** @@name Macros for the `once` condition which ignores all subsequent log
+ * calls except the first one.
+ */
+///@@{
+/**
+ * \def RCL_LOG_CONDITION_ONCE_BEFORE
+ * A macro initializing and checking the `once` condition.
+ */
+#define RCL_LOG_CONDITION_ONCE_BEFORE \
+  { \
+    static int __rcl_logging_once = 0; \
+    if (RCL_UNLIKELY(0 == __rcl_logging_once)) { \
+      __rcl_logging_once = 1;
+/**
+ * \def RCL_LOG_CONDITION_ONCE_AFTER
+ * A macro finalizing the `once` condition.
+ */
+#define RCL_LOG_CONDITION_ONCE_AFTER } \
+  }
+///@@}
+
+/** @@name Macros for the `expression` condition which ignores the log calls
+ * when the expression evaluates to false.
+ */
+///@@{
+/**
+ * \def RCL_LOG_CONDITION_EXPRESSION_BEFORE
+ * A macro checking the `expression` condition.
+ */
+#define RCL_LOG_CONDITION_EXPRESSION_BEFORE(expression) \
+  if (expression) {
+/**
+ * \def RCL_LOG_CONDITION_EXPRESSION_AFTER
+ * A macro finalizing the `expression` condition.
+ */
+#define RCL_LOG_CONDITION_EXPRESSION_AFTER }
+///@@}
+
+/** @@name Macros for the `function` condition which ignores the log calls
+ * when the function returns false.
+ */
+///@@{
+/// The filter function signature.
+/**
+ * \return true to log the message, false to ignore the message
+ */
+typedef bool (* RclLogFilter)();
```

Comment:
Additional conditions for filters with options are certainly possible. For this implementation I will stick with the simple case for now until we have a need to extend it. Additional conditions can also easily be added.